Userland/Libraries/LibGfx/ImageFormats/JBIG2Loader.cpp

```diff
@@ -4,23 +4,120 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <AK/Debug.h>
 #include <LibGfx/ImageFormats/JBIG2Loader.h>
 
 // Spec: ITU-T_T_88__08_2018.pdf in the zip file here:
 // https://www.itu.int/rec/T-REC-T.88-201808-I
 
 namespace Gfx {
 
+// JBIG2 spec, Annex D, D.4.1 ID string
+static constexpr u8 id_string[] = { 0x97, 0x4A, 0x42, 0x32, 0x0D, 0x0A, 0x1A, 0x0A };
+
+// Annex D
+enum class Organization {
+    // D.1 Sequential organization
+    Sequential,
+
+    // D.2 Random-access organization
+    RandomAccess,
+
+    // D.3 Embedded organization
+    Embedded,
+};
+
+struct JBIG2LoadingContext {
+    enum State {
+        NotDecoded = 0,
+        Error,
+    };
+    State state { State::NotDecoded };
+    ReadonlyBytes data;
+
+    Organization organization { Organization::Sequential };
+    IntSize size;
+
+    Optional<u32> number_of_pages;
+};
+
+static bool decode_jbig2_header(JBIG2LoadingContext& context)
+{
+    if (!JBIG2ImageDecoderPlugin::sniff(context.data)) {
+        dbgln_if(JBIG2_DEBUG, "Invalid JBIG2 header");
+        context.state = JBIG2LoadingContext::State::Error;
+        return false;
+    }
+
+    // D.4.2 File header flags
+    if (context.data.size() < sizeof(id_string) + sizeof(u8)) {
+        dbgln_if(JBIG2_DEBUG, "JBIG2LoadingContext: Not enough data for header");
+        context.state = JBIG2LoadingContext::State::Error;
+        return false;
+    }
+    u8 header_flags = context.data[8];
+    if (header_flags & 0b11110000) {
+        dbgln_if(JBIG2_DEBUG, "JBIG2LoadingContext: Invalid header flags");
+        context.state = JBIG2LoadingContext::State::Error;
+        return false;
+    }
+    context.organization = (header_flags & 1) ? Organization::Sequential : Organization::RandomAccess;
+    bool has_known_number_of_pages = (header_flags & 2) ? false : true;
+    bool uses_templates_with_12_AT_pixels = (header_flags & 4) ? true : false;
+    bool contains_colored_region_segments = (header_flags & 8) ? true : false;
+
+    // FIXME: Do something with these?
+    (void)uses_templates_with_12_AT_pixels;
+    (void)contains_colored_region_segments;
+
+    // D.4.3 Number of pages
+    if (has_known_number_of_pages) {
+        if (context.data.size() < sizeof(id_string) + sizeof(u8) + sizeof(u32)) {
+            dbgln_if(JBIG2_DEBUG, "JBIG2LoadingContext: Not enough data for number of pages");
+            context.state = JBIG2LoadingContext::State::Error;
+            return false;
+        }
+        context.number_of_pages = *(BigEndian<u32> const*)(&context.data[9]);
```

Comment:
Have you considered using a `Stream`?
This whole block could be replaced by `context.number_of_pages = context.stream->read<BigEndian<u32>>()` which I would consider more readable.